Add UIView.separator extension with tests and changelog 🚀

SsangG77 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Checklist
<!--- Please go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I checked the [**Contributing Guidelines**](https://github.com/SwifterSwift/SwifterSwift/blob/master/CONTRIBUTING.md) before creating this request.
- [x] New extensions are written in Swift 5.6.
- [x] New extensions support iOS 12.0+ / tvOS 12.0+ / macOS 10.13+ / watchOS 4.0+, or use `@available` if not.
- [x] I have added tests for new extensions, and they passed.
- [x] All extensions have a **clear** comments explaining their functionality, all parameters and return type in English.
- [x] All extensions are declared as **public**.
- [x] I have added a [changelog](https://github.com/SwifterSwift/SwifterSwift/blob/master/CHANGELOG_GUIDELINES.md) entry describing my changes.


🚀 This PR adds a new extension to `UIView`:

### ✅ What's added
- `separator(color:height:spacing:)` method, which makes it easy to add a bottom border/separator to any UIView.
- Fully tested with `XCTest` (`UIViewExtensionsTests`)
- `CHANGELOG.md` updated under `[Unreleased]`

### 💡 Usage

```swift
let view = UIView()
view.separator(color: .gray, height: 1, spacing: 4)
```

SsangG77 commented:
@LucianoPAlmeida 
Hi, I’d appreciate it if someone could review this PR when you have time.  
Let me know if there’s anything I should change. Thanks!

SsangG77 commented on CHANGELOG.md:
```diff
@@ -2,6 +2,14 @@
+### Added
+
+- **UIView**
+  - Added `separator(color:height:spacing:)` method to easily insert a bottom line to a view. 🚀
```
Hi @LucianoPAlmeida, I’ve updated the CHANGELOG as per your suggestion and included the PR link and username. Let me know if there’s anything else I should adjust. Thanks again for the review!

guykogus commented on CHANGELOG.md:
```diff
@@ -2,6 +2,14 @@
 
 The changelog for **SwifterSwift**. Also see the [releases](https://github.com/SwifterSwift/SwifterSwift/releases) on GitHub.
 
+## [Unreleased]
```
You don't need this new section, you already have the "Upcoming Release"

guykogus commented on Sources/SwifterSwift/UIKit/UIViewExtensions.swift:
```diff
@@ -751,4 +751,32 @@ public extension UIView {
+    ///   - color: The separator color. Default is `.black`.
+    ///   - height: The height of the separator. Default is `1`.
+    ///   - spacing: Spacing from the bottom edge. Default is `0`.
+    func separator(color: UIColor = .black, height: CGFloat = 1, spacing: CGFloat = 0) {
```
The function name should be clearer about its purpose.
```suggestion
    func addBottomSeparator(color: UIColor = .black, height: CGFloat = 1, spacing: CGFloat = 0) {
```

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,43 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
         XCTAssertFalse(view.subviews.first is UIVisualEffectView)
     }
+    
+    func testSeparatorAddsSubview() {
```
The standard for this project is that each extension function has only 1 test function.

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,43 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+            let view = UIView()
+            view.separator()
+
+            guard let separator = view.subviews.first else {
+                XCTFail("Separator not found")
+                return
+            }
```
You need to mark this func as `throws` to make this work.
```suggestion
            let separator = try XCTUnwrap(view.subviews.first, "Separator not found")
```

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,43 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+                return
+            }
+
+            XCTAssertFalse(separator.translatesAutoresizingMaskIntoConstraints)
```
This isn't very useful, as you're testing internal implementation details. It's not a requirement of the function's API that this result is exected.

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,43 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+
+            view.separator(color: expectedColor, height: expectedHeight)
+
+            guard let separator = view.subviews.first else {
+                XCTFail("Separator not found")
+                return
+            }
```
```suggestion
            let separator = try XCTUnwrap(view.subviews.first, "Separator not found")
```

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,43 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+
+            XCTAssertEqual(separator.backgroundColor, expectedColor)
+
+            let heightConstraint = separator.constraints.first { $0.firstAttribute == .height }
+            XCTAssertEqual(heightConstraint?.constant, expectedHeight)
```
Similar to the comment below - you're testing an implementation detail. Better to check something like `separator.bounds.height`.

SsangG77 commented:
@guykogus 

Thanks a lot for the detailed feedback!

Renamed the function to addViewSeparator for clarity

Made cornerRadius height-based (height / 2)

Consolidated test cases into a single method

Replaced guard let with XCTUnwrap

Removed internal implementation detail checks from tests

Cleaned up the CHANGELOG section as per the project format

Let me know if there’s anything else I should improve!

guykogus commented on CHANGELOG.md:
```diff
@@ -2,7 +2,15 @@
 
 The changelog for **SwifterSwift**. Also see the [releases](https://github.com/SwifterSwift/SwifterSwift/releases) on GitHub.
 
+
+
 ## Upcoming Release
+
+### Added
+
+- **UIView**
+  - Added `separator(color:height:spacing:)` method to easily insert a bottom line to a view. 🚀 [#1234](https://github.com/SwifterSwift/SwifterSwift/pull/1234) by [sangjin](https://github.com/SsangG77)
+
 ### Added
```
Please look at this carefully and keep everything organised neatly.
```suggestion
## Upcoming Release

### Added
- **UIView**
  - Added `separator(color:height:spacing:)` method to easily insert a bottom line to a view. 🚀 [#1234](https://github.com/SwifterSwift/SwifterSwift/pull/1234) by [sangjin](https://github.com/SsangG77)
```

guykogus commented on Sources/SwifterSwift/UIKit/UIViewExtensions.swift:
```diff
@@ -751,4 +751,32 @@ public extension UIView {
+    ///   - color: The separator color. Default is `.black`.
+    ///   - height: The height of the separator. Default is `1`.
+    ///   - spacing: Spacing from the bottom edge. Default is `0`.
+    func addViewSeparator(color: UIColor = .black, height: CGFloat = 1, spacing: CGFloat = 0) {
```
Please use the name I suggested before. It makes it clear to the caller that the separator will be added to the bottom of the view.
```suggestion
    func addBottomSeparator(color: UIColor = .black, height: CGFloat = 1, spacing: CGFloat = 0) {
```

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,24 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+    }
+
+    
+    
```
There are excessive empty spaces here. Do you have `swiftformat` and `swiftlint` installed? If not, please install them to make sure that the code meets the project standards.

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,24 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+        
+        XCTAssertEqual(view.subviews.count, 1)
+        XCTAssertEqual(separator.backgroundColor, expectedColor)
+        XCTAssertEqual(separator.frame.height, expectedHeight, accuracy: 0.1)
```
Why do you need `accuracy`? It should be exact.

SsangG77 commented:
Hi @guykogus 

Thanks for the feedback! I've made the requested changes:

Renamed the method to addBottomSeparator for clarity
Removed the accuracy from the test as the height should match exactly
Cleaned up extra empty lines in the test file
Updated the CHANGELOG.md to match the formatting conventions

Let me know if there's anything else you'd like me to adjust!

guykogus commented:
Hi @SsangG77, I think you forgot to push your changes.

SsangG77 commented:
@guykogus
Oh sorry! I pushed now.

guykogus commented on CHANGELOG.md:
```diff
@@ -2,7 +2,14 @@
 
 The changelog for **SwifterSwift**. Also see the [releases](https://github.com/SwifterSwift/SwifterSwift/releases) on GitHub.
 
+
+
```
Please remove these extra empty lines

guykogus commented on CHANGELOG.md:
```diff
@@ -2,7 +2,14 @@
+
 ## Upcoming Release
+
+### Added
+- **UIView**
+  - Added `addBottomSeparator(color:height:spacing:)` method to easily insert a bottom line to a view. 🚀 [#1234](https://github.com/SwifterSwift/SwifterSwift/pull/1234) by [sangjin](https://github.com/SsangG77)
+
 ### Added
 - **Android**
```
You have `Added` twice
```suggestion
### Added
- **UIView**
  - Added `addBottomSeparator(color:height:spacing:)` method to easily insert a bottom line to a view. 🚀 [#1234](https://github.com/SwifterSwift/SwifterSwift/pull/1234) by [sangjin](https://github.com/SsangG77)
- **Android**
```

guykogus commented on Sources/SwifterSwift/UIKit/UIViewExtensions.swift:
```diff
@@ -751,4 +751,32 @@ public extension UIView {
 
+
+
+public extension UIView {
```
Why have you added a new extension section? Can you move this into the right section? Maybe at line 703?

guykogus commented on Sources/SwifterSwift/UIKit/UIViewExtensions.swift:
```diff
@@ -751,4 +751,32 @@ public extension UIView {
     }
 }
 
+
+
```
Please take note of all the extra empty lines you're adding. Look at the rest of the file/s and make sure that you're aligning with the project style.

guykogus commented on Tests/UIKitTests/UIViewExtensionsTests.swift:
```diff
@@ -683,6 +683,20 @@ final class UIViewExtensionsTests: XCTestCase { // swiftlint:disable:this type_b
+    
+    func testAddBottomSeparator() throws {
+        let view = UIView()
+        let expectedColor: UIColor = .red
```
```suggestion
        let expectedColor = UIColor.red
```

SsangG77 commented:
@guykogus 
Thanks for the feedback, I pushed it.

guykogus commented on CHANGELOG.md:
```diff
@@ -3,7 +3,11 @@
 The changelog for **SwifterSwift**. Also see the [releases](https://github.com/SwifterSwift/SwifterSwift/releases) on GitHub.
 
 ## Upcoming Release
+
```
As I mentioned before, look at the current style and match it. You've added a new empty line here.

guykogus commented on CHANGELOG.md:
```diff
@@ -3,7 +3,11 @@
 ### Added
+- **UIView**
+  - Added `addBottomSeparator(color:height:spacing:)` method to easily insert a bottom line to a view. 🚀 [#1234](https://github.com/SwifterSwift/SwifterSwift/pull/1234) by [sangjin](https://github.com/SsangG77)
+
```
You've also added a new empty line here.

guykogus commented on Sources/SwifterSwift/UIKit/UIViewExtensions.swift:
```diff
@@ -700,6 +700,29 @@ public extension UIView {
+        self.addSubview(line)
+        NSLayoutConstraint.activate([
+            line.heightAnchor.constraint(equalToConstant: height),
+            line.leadingAnchor.constraint(equalTo: self.leadingAnchor),
+            line.trailingAnchor.constraint(equalTo: self.trailingAnchor),
```
I don't think you need `self.` here. `swiftformat` should've taken care of that. Do you have it installed?
```suggestion
            line.leadingAnchor.constraint(equalTo: leadingAnchor),
            line.trailingAnchor.constraint(equalTo: trailingAnchor),
```

SsangG77 commented:
@guykogus 
Thanks! All requested changes have been made. Let me know if there's anything else you'd like me to update.

guykogus reviewed (approved):
Great. Thanks for your patience with the back and forth :)

SsangG77 commented:
Thanks for the review and approval! 

Just to note:
- The test coverage warning is due to the new method not being covered. I can add a test if needed.
- The Android CI failure seems unrelated as this is an iOS-only UIKit extension.

Let me know if there's anything else I should address!
